File: clojure/clj-scratch/src/clj_scratch/async.clj
(ns clj-scratch.async
  (use [clojure.core.async]))

(let [c (chan)]
  (go
   (Thread/sleep (rand-int 1000))
   (>! c "success!"))

  (<!!
   (go
    (let [[result source] (alts! [c (timeout 500)])]
      (if (= source c)
        (println "got a value")
        (println "timeout"))))))

File: go/async.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func main() {
	rand.Seed(time.Now().UTC().UnixNano())

	c := make(chan string)

	go func() {
		sleep := time.Duration(rand.Intn(1000))
		time.Sleep(sleep * time.Millisecond)
		c <- "success!"
	}()

	select {
		case <-c:
			fmt.Println("Got a value!")
		case <- time.After(500 * time.Millisecond):
			fmt.Println("Timeout")
	}
}
